File: test-suite/handwritten-src/ts/TokenTest.ts
import {TestCase, allTests, assertTrue, assertSame} from "./testutils"
import * as test from "../../generated-src/ts/test";

class JsToken {
    whoami() {
        return "Javascript";
    }
}

export class TokenTest extends TestCase {
    m: test.Test_statics;

    constructor(module: test.Test_statics) {
        super(module);
        this.m = module;
    }
    
    testTokens() {
        var jt = new JsToken();
        assertSame(this.m.testsuite.TestHelpers.tokenId(jt), jt);
    }

    testNullToken() {
        assertSame(this.m.testsuite.TestHelpers.tokenId(null), null);
    }

    testCppToken() {
        var ct = this.m.testsuite.TestHelpers.createCppToken();
        assertSame(this.m.testsuite.TestHelpers.tokenId(ct), ct);
        this.m.testsuite.TestHelpers.checkCppToken(ct);
        ct = null;
    }

    testTokenType() {
        this.m.testsuite.TestHelpers.checkTokenType(new JsToken(), "Javascript");
        this.m.testsuite.TestHelpers.checkTokenType(this.m.testsuite.TestHelpers.createCppToken(), "C++");
        var threw = false;
        try {
            this.m.testsuite.TestHelpers.checkTokenType(new JsToken(), "foo");
        } catch (e) {
            threw = true;
        }
        assertTrue(threw);
        threw = false;
        try {
            this.m.testsuite.TestHelpers.checkTokenType(this.m.testsuite.TestHelpers.createCppToken(), "foo");
        } catch (e) {
            threw = true;
        }
        assertTrue(threw);
    }

    testNotCppToken() {
        var threw = false;
        try {
            this.m.testsuite.TestHelpers.checkCppToken(new JsToken());
        } catch (e) {
            threw = true;
        }
        assertTrue(threw);
    }
}

allTests.push(TokenTest);

File: test-suite/handwritten-src/swift/TokenTest.swift
import XCTest
@testable import TestSuite

final class TokenTest: XCTestCase {

    class SwiftToken: UserToken {
        func whoami() throws -> String { return "Swift" }
    }

    func testTokens() throws {
        let st = SwiftToken()
        XCTAssertIdentical(try TestHelpers_statics.tokenId(st), st)
    }

    func testNullToken() throws {
        XCTAssertNil(try TestHelpers_statics.tokenId(nil))
    }

    func testCppToken() throws {
        let ct = try TestHelpers_statics.createCppToken();
        XCTAssertIdentical(try TestHelpers_statics.tokenId(ct), ct)
        try TestHelpers_statics.checkCppToken(ct)
    }

    func testTokenType() throws {
        try TestHelpers_statics.checkTokenType(SwiftToken(), type:"Swift")
        try TestHelpers_statics.checkTokenType(TestHelpers_statics.createCppToken(), type:"C++");
        var threw = false
        do {
            try TestHelpers_statics.checkTokenType(SwiftToken(), type:"foo")
        } catch {
            threw = true
        }
        XCTAssertTrue(threw)
    }

    func testNotCppToken() {
        var threw = false
        do {
            try TestHelpers_statics.checkCppToken(SwiftToken())
        } catch {
            threw = true
        }
        XCTAssertTrue(threw)
    }
}
